### PR TITLE
[SPARK-47456][SQL] Support ORC Brotli codec

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
This PR aims to support ORC Brotli codec.

### Why are the changes needed?
Currently, the master branch of Spark has used ORC 2.0.0 version and supports brotli compression encoding. ([SPARK-44115](https://issues.apache.org/jira/browse/SPARK-44115))
However, due to some hardcode checks in Spark, brotli cannot be used normally.

```sql
set spark.sql.orc.compression.codec=BROTLI;
```

```
java.lang.IllegalArgumentException: The value of spark.sql.orc.compression.codec should be one of uncompressed, lz4, lzo, snappy, zlib, none, zstd, but was brotli
```

See https://github.com/apache/orc/pull/1714

### Does this PR introduce _any_ user-facing change?
Yes

In this PR, add the corresponding brotli jar package, we can use orc brotli encoding.

```xml
    <dependency>
      <groupId>com.aayushatharva.brotli4j</groupId>
      <artifactId>brotli4j</artifactId>
    </dependency>
```

REF: https://github.com/netty/netty/blob/3cd364107167600e8eb4b0b85553ed895519e2ed/codec/pom.xml#L91-L125

### How was this patch tested?
local test

image

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by yaooqinn — APPROVED** [review]
LGTM from my side, also cc @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala [hunk 1]
```diff
@@ -71,6 +71,9 @@ class OrcCodecSuite extends FileSourceCodecSuite {
 
   override def format: String = "orc"
   override val codecConfigName: String = SQLConf.ORC_COMPRESSION.key
+  // Exclude "BROTLI" because its dependencies
+  // require adding different jars according to different OSs
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:75]: Which jar should be used when the user wants to enable this feature?

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:75]: This one. I excluded intentionally during upgrading to Apache ORC 2.0.0.

https://github.com/apache/spark/blob/681b41f0808e35313f4b63ca735f604d9ffc687f/pom.xml#L2615-L2618

**Review by dongjoon-hyun — APPROVED** [review]

File: docs/sql-data-sources-orc.md [hunk 2]
```diff
@@ -241,7 +241,7 @@ Data source options of ORC can be set via:
   <tr>
     <td><code>compression</code></td>
     <td><code>zstd</code></td>
-    <td>compression codec to use when saving to file. This can be one of the known case-insensitive shorten names (none, snappy, zlib, lzo, zstd and lz4). This will override <code>orc.compress</code> and <code>spark.sql.orc.compression.codec</code>.</td>
+    <td>compression codec to use when saving to file. This can be one of the known case-insensitive shorten names (none, snappy, zlib, lzo, zstd, lz4 and brotli). This will override <code>orc.compress</code> and <code>spark.sql.orc.compression.codec</code>.</td>
```
Comment (dongjoon-hyun) [inline docs/sql-data-sources-orc.md:244]: Like Parquet, we need a `Note` here. 

https://github.com/apache/spark/blob/5f48931fcdf712c3646a368ae9643b5ccba00db2/docs/sql-data-sources-parquet.md?plain=1#L488

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:75]: As he mentioned, native brotli should be available in the system, too.

**Review by LuciferYang — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:75]: OK~

Comment (cxzl25) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:75]: Thanks for the comment, I've added the dependencies that need to be added to the PR description.

File: docs/sql-data-sources-orc.md (hunk 2 above)
Comment (cxzl25) [reply docs/sql-data-sources-orc.md]: Thanks. Updated.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

Thank you, @cxzl25 , @yaooqinn , @LuciferYang !

**Pull request closed without merging**